Improve universe generation performance.

LGM-Doyle opened the pull request:
This PR along with #1214 provide a performance improvement to universe generation.

The performance improvement on a galaxy with 999 systems and 39 AIs is from 83 seconds to to 21 seconds.

39 seconds of that improvement is from #1214, which is linux only, but I don't think that windows machines are experiencing that slowdown.

This PR makes the improvements by making the following changes in the universe generation python code along with supporting changes in the C++ code:  
- It vectorizes `locations()` and `special_locations()`.  By searching all locations at the same time, it reduces round trip time between python and C++ and does all of the `WithinStarLaneJumps` calculations in parallel.
- It aggressively trims the candidate list for home systems.
- It rearranges the HONEYCOMB condition to do the expensive clause last.
- It uses known home systems to remove systems from consideration for monsters and natives.

While doing this two bugs were found, but only one was fixed.

In monsters and specials the systems were not randomized.  In galaxy topologies were system number is strongly correlated with location the monsters and specials ended up bunched in one area of the galaxy.  This is fixed in this PR.

In `specials.py` specials check their own clause against all existing specials before placement, but they do not recursively check all existing specials clauses against themselves.  Since, some special's clauses are written in a reciprocal way this is not a problem for all specials.  I could not think of a computationally efficient way to fix this so I made a TODO note and did not fix it.

LGM-Doyle committed:
Shuffle systems in monsters.py and specials.py

Shuffling systems prevents universe_obj id dependent placement.  In
particular placement limits meant that specials with limits were more
likely to be placed in the earlier systems.

LGM-Doyle committed:
Rearrange HONEYCOMB condition so that the expensive clause is the last AND term

LGM-Doyle committed:
Expose SpecialLocations() instead of SpecialLocation() in ServerWrapper

Searching all of the locations at the same time during specials
placement is faster because the WithinStarLaneJumps calculation is only
done once.

Refactor Locations() from MonsterFleetPlanWrapper to use a common
function, FilterIDSWithCondition().

LGM-Doyle committed:
Move get_systems_within_jumps() to galaxy_topology.py

Move get_systems_within_jumps() to a new file galaxy_topology.py.
galaxy_topology.py will store common utility functions after starlanes
have been added to the galaxy   This anticipates and fixes cyclic import
problems with statistics.py, importing natives.py importing empires.py,
importing statistics.py

LGM-Doyle committed:
Use home systems to filter systems tried for placement in natives.py

LGM-Doyle committed:
Refactor specials.py to improve performance speed.

Apply base_chance to specials up front to remove objects that will not
have specials from the pool.

Define a GALAXY_DECOUPLING_DISTANCE which is the largest distance over
which specials interact with anything.

Consider groups of objects that are mutually farther apart than
GALAXY_DECOUPLING_DISTANCE so that their spawning rates can be
calculated collectively and quickly.

LGM-Doyle committed:
Expose Locations() instead of Location() in MonsterFleetPlanWrapper.

Searching all of the locations at the same time during monster placement
is faster because the WithinStarLaneJumps calculation is only done once.

LGM-Doyle committed:
Improve find_home_systems_for_min_jump_distance empires.py.

Speed up find_home_systems_for_min_jump_distance() used during home
system selection during galaxy generation, by aggressively trimming the
candidate list of systems.

An example improvement for 1000 systems with 39 empires is from 11 s to
230 ms.

LGM-Doyle committed:
Replace special_location() with special_locations() in monsters.py

Using special_locations on a list of planets to compute nest locations
improves performance of monster location generation.
For example in a 1000 system galaxy generate_monsters run time improves
from 2.2 s to 1.1 s

LGM-Doyle committed:
Use home systems to filter systems tried for placement in monsters.py

LGM-Doyle committed:
Assorted PEP8 fixes within universe generation.

1. Move standard 'import statistics' before 'import freeorion as fo'
2. Line length
3. Change checks from 'x == True' to 'x is True'
4. Fix inner scope variable e shadowing outer scope e.
5. White around arithmetic operators.

Morlic-fo reviewed (changes requested):
The special refactoring does not look right to me. Will have to look into it in more detail though.

Cjkjvfnby reviewed (commented):
I make a quick look.  I will not add `Request changes` since does not know if I will have time to check this RP again.

LGM-Doyle committed:
Improved comments for Review #1225

LGM-Doyle committed:
Code changes in empires.py for review #1225.

1. Changed variable names from best_case_worst_merit to
best_case_lower_bound and current_worst_merit to
current_merit_lower_bound.
2. Locally cull the systems_pool in HoemSystemFinder.
3. Use python set differencing.
4. Reduce nesting in early exit statement.

LGM-Doyle committed:
In empires.py HomeSystemFinder more aggresively reducing number of iterations.

Make sure that number of attempts is never greater than the remaining
number of systems in the local pool.

LGM-Doyle committed:
Code changes in monsters.py and natives.py for review #1225.

1. Remove unnecessary list([]) bracketing of from_iterable.
2. Remove vacuous test in calculation of fl_location_cache length.
3. Improve calculation of tracked_nest_valid_locations.

LGM-Doyle committed:
Code changes in specials.py for review #1225.

1. Add missing and rare 4 special locations.
2. Rename place_special_fast to place_special.
3. Use format for print statement.
4. Name tuple elements.

LGM-Doyle reviewed (commented):
Thanks for the thorough review.

LGM-Doyle committed:
Fix handling of @content_tag in monsters.py from review #1225.

The entire contents of @content_tag needs to be on one line regardless
of length.

Vezzra reviewed (changes requested):
You seem to have a soft spot for big PRs, don't you? 😉 That biggie could probably have been broken up into a couple of smaller PRs.

Well, this way I can't review the entire thing at once, so I'll have to break that up into several smaller reviews. Can't do all of it today, so more to come.

This first one deals with the changes to `empires.py` and splitting `get_systems_within_jumps` from `galaxy.py` to its own module.

LGM-Doyle committed:
In empires.py move attempts decrement to where len(local_pool) is non-zero.

This issue was raised in the review #1225.

LGM-Doyle committed:
In empires.py store merit with system in local list, to avoid recalculation.

LGM-Doyle committed:
In empires.py add report of failure when local_pool is too short.

LGM-Doyle committed:
In empires.py fix sort order to agree with algorithm description.

This error was reported in review #1225.

LGM-Doyle committed:
In empires.py tighten the exit conditions in HomeSystemFinder.

Fixing the sort order bug found in review #1225 greatly increased the
number and time used in unsuccessful searches for home systems.

This commit adds/tightens 4 early exit conditions:
1. Adds exit early if the minimum limit is achieved.
2. When a succesful candidate is found increase the merit lower bound
one greater than the successful candidate so that the next good
candidate is better.
3. If repeated candidate sets are much smaller than the target, then
exit early since this means that either the min jumps spacing is too
large or the systems pool is too small.
4. If candidate set is close to the correct size, then keep trying.

LGM-Doyle committed:
In empires.py remove variable best_case_merit_lower_bound.

That exit condition no longer saves an iteration through the loop.

LGM-Doyle committed:
In specials.py rename calc_num_placed to calculate_number_of_specials_placed.

As requested in review #1225.

LGM-Doyle committed:
In specials.py make place_special return nothing as it always places a special.

From review #1225.

LGM-Doyle committed:
In specials.py rename variables.

sys -> system, because of conflicts with global module sys.
n -> specials_count

LGM-Doyle committed:
In specials.py remove timers.

LGM-Doyle committed:
In unverse_generation restore import location of statistics.

The universe_generation module statistics shadows the standard library
module statistics and its import generates a spurious PEP8 warning to
place it with the standard import.

LGM-Doyle committed:
In specials.py prevent division by zero.

Based on review #1225 comment.

LGM-Doyle committed:
Minor fixes from review #1225.

1. Improve docstring.
2. Remove unnecessary str() in debug message.

LGM-Doyle reviewed (commented):
Things I missed in the last pass that I've fixed.
1.  Explicitly prevent division by zero in place_special.
2. Improve place_special docstring
3. Remove unneeded str() in debug message.

Vezzra reviewed (changes requested):
Second pass on the new home system finder algorithm. Two issues remain. Otherwise looks good.

Will follow up with separate reviews for the changes to `monsters.py` and `specials.py`.

Vezzra reviewed (changes requested):
Review of the revised monster distribution.

Vezzra reviewed (changes requested):
Review for changes to `ServerWrapper.cpp`.

LGM-Doyle committed:
Restore current_merit_lower_bound to be the current best merit.

There is no push it by one it already excludes these systems in the
filtering.

Issue raised during review #1225.

LGM-Doyle committed:
In empires.py move early successful exit to after candidate is updated.

Issue raised during review #1225.

LGM-Doyle committed:
In empires.py refactor home system finder exit due to impossible problem.

Compare the mean and standard deviation of the lengths of the candidate
home systems against the target number of home systems and quit early
when unable to succeed.

LGM-Doyle committed:
In empires.py home system finder simplify exit when no candidate was found.

LGM-Doyle committed:
In empires.py clarify distinction between merit and lowest merit in a set.

From concerns in review #1225.

LGM-Doyle committed:
Housekeeping: operator spacing in galaxy.py.

From review #1225.

LGM-Doyle committed:
In empires.pt remove near empire filtering for monsters.

As pointed out in review #1225 there is no introspection to determine if
a minimum distance from home systems to monsters exists.

LGM-Doyle committed:
In ServerWrapper.cpp improve comments.  From review #1225.

LGM-Doyle committed:
Housekeeping:  spelling.

LGM-Doyle committed:
Use C++11 range based for loops in ServerWrapper.cpp::FilterIDsWithCondition().

LGM-Doyle reviewed (commented):
I also addressed the following concern, but couldn't find the original request to respond directly.

@Morlic-fo I have changed phrasing to lowest merit in a set to avoid talking about the merit of a set.

Morlic-fo reviewed (commented):
I am not convinced that the new heuristic for early exit is valid.

LGM-Doyle committed:
In empires.py change reverse to descending, for review #1225.

LGM-Doyle committed:
In empires.py change merit check to > from >= so that merit only improves.

From review #1225.

LGM-Doyle committed:
In empires.py fix the variance based exit condition.

1. It should scale with the sqrt of the number of systems.
2. It should use the number of systems, not the number of systems in the
pool.

See discussion in review #1225.

Vezzra reviewed (changes requested):
Ok, *finally* got around to review the changes to the specials distribution algorithm (sorry it took me so long). Aside from my comments looks good - as far as I can tell and have been able to wrap my brain around it. :wink:

However, as stated in my comments, this can only be a stop-gap solution, given the current compromises wrt to correctness.

LGM-Doyle committed:
Adjust low candidate threshold exit condition for review #1225.

LGM-Doyle committed:
Minor fixes for review #1225 in specials.py.

1. Change _ to __.
2. Move comment.
3. Remove debugging variable i_cache.

LGM-Doyle committed:
Remove SpecialLocation().

This should have been removed when SpecialLocations() was added.

LGM-Doyle committed:
Prevent OSX/Clang list based initialization of vector in HumanClientApp.cpp.

LGM-Doyle reviewed (commented):
I rebased to master to pick up the mentioned bugfix commit.

I added a commit which might fix your compile problem, which I think is from Clang trying to use the list initializer for the vector, when it won't work.  I don't know why this isn't also a problem in master.

I'm not suggesting sacrificing correctness for speed.  The recursive placement bug has existed for a least a year.  No one noticed.  It is fairly benign.  When I prepared this PR, I figured that it is fast and no more broken than the existing code.

Startup time is somewhat important for three reasons:
- Perception of the game quality.  If it takes 10 minutes to generate a 100 system game on a 1GHz machine, is it reasonable to take 600e6 operations per system?  A good place to start improving game performance is at the beginning.
- Testing.  I wrote this code when I was testing something and each test case initially took 3 minutes each.  I hacked it down to 14 seconds and was able to generate many test cases.  Yah.  I hope someone else improves it further to 1 second.  
- Extensibility of the software.  If someone proposes a new feature that will double startup time, the threshold for accepting that change is different at a 10 second startup vs. 10 minutes startup time.      

Speaking of doubling startup time.  I have an extension to this PR, that is not complete yet.  It may double statrup time, but it will correct the recursive location condition problem.

LGM-Doyle committed:
Minor fixes for review #1225.

1. Remove unused variable len_systems.
2. Capitalize constant MISS_THRESHOLD.
